### PR TITLE
FEATURE: added liquid based templating

### PR DESCRIPTION
**Pull request by fzngagan**

### CODE REVIEW

File: spec/components/custom_wizard/mapper_spec.rb [hunk 1]
```diff
@@ -41,6 +41,30 @@
...
     ).read)
   }
+  let(:template_params) {
+    {
+      "step_1_field_1" => "Hello"
+    }
+  }
+  let(:template_params_non_empty) {
+    {
+      "step_1_field_1" => nil,
```
Comment (angusmcleod) [inline spec/components/custom_wizard/mapper_spec.rb:51]: @fzngagan shouldn't be using arrows in hash literals. Use a colon instead.

File: spec/components/custom_wizard/mapper_spec.rb [hunk 2]
```diff
@@ -250,4 +274,75 @@
...
+      result = template_mapper_non_empty.interpolate(
+        template.dup,
+        template: true,
+        user: true,
+        wizard: true,
+        value: true
+      )
+      expect(result).to eq(template_params_non_empty["step_1_field_3"])
+    end
+  end
```
Comment (angusmcleod) [inline spec/components/custom_wizard/mapper_spec.rb:347]: @fzngagan We're missing a spec that explicitly tests the case we discussed

```
{% if "step_1_field_{1|2|3|4|5}" %}
  {%= "step_1_field_{1|2|3|4|5}" %}
{% end %}
```

File: spec/components/custom_wizard/mapper_spec.rb [hunk 3]
```diff
@@ -250,4 +274,75 @@
       user: user1
     ).perform).to eq(false)
   end
+
+  context "Output Templating" do
```
Comment (angusmcleod) [inline spec/components/custom_wizard/mapper_spec.rb:278]: @fzngagan We need a test for the case when ``template: false``.

File: spec/components/custom_wizard/mapper_spec.rb [hunk 4]
```diff
@@ -250,4 +274,75 @@
...
+        template.dup,
+        template: true,
+        user: true,
+        wizard: true,
+        value: true
+      )
+      expect(result).to eq(template_params["step_1_field_1"].size.to_s)
+    end
+
+    it "custom filter: 'first_non_empty' gives first non empty element from list" do
```
Comment (angusmcleod) [inline spec/components/custom_wizard/mapper_spec.rb:333]: @fzngagan We need a test for when all elements are empty. Also for when multiple elements are not empty. Currently we're making the test too easy, i.e. there's only one non-empty and it's at the end.

**Review by angusmcleod — CHANGES_REQUESTED** [review]

File: spec/components/custom_wizard/mapper_spec.rb (hunk 2 above)
Comment (fzngagan) [reply spec/components/custom_wizard/mapper_spec.rb:402]: To simplify this, I've create a liquid filter `first_non_empty` :)

File: spec/components/custom_wizard/mapper_spec.rb (hunk 3 above)
Comment (fzngagan) [reply spec/components/custom_wizard/mapper_spec.rb:280]: I've added a test for this

File: spec/components/custom_wizard/mapper_spec.rb (hunk 4 above)
Comment (fzngagan) [reply spec/components/custom_wizard/mapper_spec.rb]: Added a few more tests to ensure its doing the right thing

File: spec/components/custom_wizard/mapper_spec.rb [hunk 5]
```diff
@@ -250,4 +276,135 @@
...
+          wizard: true,
+          value: true
+        )
+        expect(result).to eq("")
+      end
+    end
+  end
+end
+
+def create_template_mapper(data, user)
```
Comment (angusmcleod) [inline spec/components/custom_wizard/mapper_spec.rb:405]: @fzngagan If you look at other examples of method definitions in our specs, they're at the start of the file

File: spec/components/custom_wizard/mapper_spec.rb [hunk 6]
```diff
@@ -250,4 +276,135 @@
...
+          template.dup,
+          template: true,
+          user: true,
+          wizard: true,
+          value: true
+        )
+        expect(result).to eq(template_params_non_empty["step_1_field_3"])
+      end
+
+      it "gives first non empty element from list  when multiple non empty values present" do
```
Comment (angusmcleod) [inline spec/components/custom_wizard/mapper_spec.rb:368]: @fzngagan There's an additional space in this line. Have you run rubocop on this?

File: spec/components/custom_wizard/mapper_spec.rb [hunk 7]
```diff
@@ -250,4 +276,135 @@
       user: user1
     ).perform).to eq(false)
   end
+
+  context "Output Templating" do
```
Comment (angusmcleod) [inline spec/components/custom_wizard/mapper_spec.rb:280]: @fzngagan "Output Templating" doesn't need to be capitalized.

**Review by angusmcleod — CHANGES_REQUESTED** [review]

File: spec/components/custom_wizard/mapper_spec.rb (hunk 6 above)
Comment (fzngagan) [reply spec/components/custom_wizard/mapper_spec.rb:368]: Are you sure. Rubocop says no offences detected. Also checked the line on the editor.

File: spec/components/custom_wizard/mapper_spec.rb (hunk 7 above)
Comment (fzngagan) [reply spec/components/custom_wizard/mapper_spec.rb]: fixed

File: spec/components/custom_wizard/mapper_spec.rb (hunk 5 above)
Comment (fzngagan) [reply spec/components/custom_wizard/mapper_spec.rb]: fixed

File: spec/components/custom_wizard/mapper_spec.rb (hunk 6 above)
Comment (angusmcleod) [reply spec/components/custom_wizard/mapper_spec.rb:375]: > from list  when multiple

should be

> from list when multiple

Comment (angusmcleod) [reply spec/components/custom_wizard/mapper_spec.rb:375]: @fzngagan You can see it just above

Screenshot at Apr 09 14-58-00

Comment (fzngagan) [reply spec/components/custom_wizard/mapper_spec.rb]: Oh I see. Rubocop would consider it an error if was a trailing whitespace at the end of the line or an indentation issue.

Fixed :)
